### PR TITLE
Feature/df 2996 add avp business case importer

### PR DESCRIPTION
**Pull request by mdavidovski**

### CODE REVIEW

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/AVPBusinessCaseImporterApplication.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.cloud.task.configuration.EnableTask;
+
+@EnableTask
+@EnableBatchProcessing
+@EnableDiscoveryClient
+@EnableFeignClients
+@EnableBinding(Source.class)
+@SpringBootApplication
+public class AVPBusinessCaseImporterApplication {
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/AVPBusinessCaseImporterApplication.java:18]: Can you please change on all classes "AVP" to "Avp", so it's consistent overall.

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import ch.admin.seco.alv.service.onlineform.dto.LamdaBusinessCase;
+import org.springframework.batch.item.database.JdbcCursorItemReader;
+import org.springframework.batch.item.database.builder.JdbcCursorItemReaderBuilder;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import javax.sql.DataSource;
+
+@Component
+public class LamdaBusinessCaseItemReaderFactory {
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java:12]: I think more appropriate name would be `StesItemReaderFactoy`

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			"       s.txk_mobile_nr                                                    as mobile," +
+			"       s.txk_email                                                        as email," +
+			"       s.txk_email                                                        as email," +
+			"       s.dat_geburts_datum                                                as birthday," +
+			"       s.txk_wohn_adresse_strasse || ' ' || s.txk_wohn_adresse_haus_nr    as street," +
+			"       plzort_stes.code                                                   as zipCode," +
+			"       plzort_stes.text_de                                                as city" +
+			" FROM stes s" +
+			"       left outer join aux_plz_v plzort_stes on s.cfa_plz_wohn_adresse = plzort_stes.id" +
+			" ORDER BY personNumber";
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java:31]: you're missing WHERE clause here ...
https://alv-ch.atlassian.net/wiki/spaces/JR2/pages/805830662/Business+Rules
> das aktuelle Datum >= “Stellenantritt am”
other part still can't be added, but this one is already visible. It's `dat_stellenantritt_ab` in STES

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/dto/AVPBusinessCaseEventDto.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
+package ch.admin.seco.alv.service.onlineform.dto;
+
+import java.time.LocalDate;
+
+public class AVPBusinessCaseEventDto {
+
+	private final String personNumber;
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/dto/AVPBusinessCaseEventDto.java:7]: why not adding here annotated validation, instead of one in constructor?

File: app-onlineform/src/main/resources/liquibase/changelog/20200511080000_added_avp_attachment.xml [hunk 5]
```diff
@@ -21,9 +21,6 @@
             <column name="OWNER_USER_ID" type="VARCHAR(36)">
                 <constraints nullable="false"/>
             </column>
-            <column name="PERSON_NUMBER" type="VARCHAR(100)">
```
Comment (emina-str) [inline app-onlineform/src/main/resources/liquibase/changelog/20200511080000_added_avp_attachment.xml:24]: you can't delete it, we need this info, I think. especially for kofax, if no where else.

File: app-onlineform/src/test/java/ch/admin/seco/alv/service/onlineform/infrastructure/web/controller/AvpRestControllerIntTest.java [hunk 6]
```diff
@@ -75,29 +86,6 @@ public void tearDown() {
         DomainEventPublisher.set(null);
     }
 
-    @Test
-    @WithSysAdmin
-    public void testCreateAvP() throws Exception {
```
Comment (emina-str) [inline app-onlineform/src/test/java/ch/admin/seco/alv/service/onlineform/infrastructure/web/controller/AvpRestControllerIntTest.java:80]: why did you delete this test?

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java [hunk 7]
```diff
@@ -63,56 +71,133 @@
...
+                        createAvpDto.getYearMonth(),
+                        createAvpDto.getOwnerUserId(),
+                        AvpAssembler.toUserInfo(createAvpDto.getUserInfo()),
+                        null)
+        );
+        publish(new AvpCreatedEvent(avp));
+    }
+
+    private void createAvpBusinessCaseForYearMonth(CreateAvpBusinessCaseDto createAvpBusinessCaseDto) {
+        AVPBusinessCase avpBusinessCase = new AVPBusinessCaseBuilder().setYearMonth(YearMonth.from(TimeMachine.now()))
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:140]: format

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/HousekeepingResponse.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package ch.admin.seco.alv.service.onlineform.application.avp;
+
+public class HousekeepingResponse {
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/HousekeepingResponse.java:3]: don't create yet another HousekeepingResponse, use the one from NpA (extract exinsting one to main package)

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/AvpDto.java [hunk 9]
```diff
@@ -15,6 +15,7 @@
 import java.util.List;
 import java.util.Set;
 
+//TODO: Check if we need ownerUserId and personNumber in frontend
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/AvpDto.java:18]: ownerUserId most definitely yes, personNumber no.

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/UserInfoDto.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
+package ch.admin.seco.alv.service.onlineform.application.avp.dto;
+
+import ch.admin.seco.alv.service.onlineform.domain.PersonNumber;
+
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import java.time.LocalDate;
+
+public class UserInfoDto {
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/UserInfoDto.java:10]: please rename to `AvpUserInfoDto`

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/npa/UserRegisteredEvent.java [hunk 11]
```diff
@@ -3,6 +3,7 @@
 import ch.admin.seco.alv.service.onlineform.domain.PersonNumber;
 import ch.admin.seco.alv.service.onlineform.domain.UserId;
 
+//TODO: Extract to common package since is used by different modules (npa, bu, avp...)
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/npa/UserRegisteredEvent.java:6]: yes

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AssemblerTest.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+import ch.admin.seco.alv.service.onlineform.application.avp.dto.CreateAvpBusinessCaseDto;
+import ch.admin.seco.alv.service.onlineform.application.avp.dto.CreateAvpDto;
+import org.junit.Test;
+
+import static ch.admin.seco.alv.service.onlineform.application.avp.AvpApplicationServiceTest.*;
+import static ch.admin.seco.alv.service.onlineform.domain.avp.AvpFixture.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AssemblerTest {
```
Comment (emina-str) [inline application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AssemblerTest.java:11]: name is too generic.

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/UserInfo.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+import javax.persistence.*;
+import javax.validation.Valid;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import java.time.LocalDate;
+
+@Embeddable
+public class UserInfo {
```
Comment (emina-str) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/UserInfo.java:13]: i think it would be better to rename it to AvpUserInfo. 
it seems really Avp specific? what do you think?

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/BusinessCaseToMessageChannelWriter.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        items.forEach(this::writeItem);
+    }
+
+    private void writeItem(AVPBusinessCaseEventDto item) {
+        this.messageChannel.send(
+            MessageBuilder.withPayload(item)
+                .setHeader(MessageHeaders.PARTITION_KEY, item.getPersonNumber())
+                .setHeader(MessageHeaders.SOURCE_SYSTEM, "avp-business-case-importer")
+                .setHeader(MessageHeaders.TARGET_SYSTEM, "online-form-service")
+                .setHeader(MessageHeaders.PAYLOAD_TYPE, item.getClass().getSimpleName())
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/BusinessCaseToMessageChannelWriter.java:29]: `.setHeader(MessageHeaders.RELEVANT_ID, item.getPersonNumber()))`
so each message can be unique, add the ID, but only PersonNumber, it's enough

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/AVPBusinessCaseImporterApplication.java (hunk 1 above)
Comment (mdavidovski) [reply app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/AVPBusinessCaseImporterApplication.java:18]: yeah, I used capitals because that's how we have it for npa

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java (hunk 3 above)
Comment (mdavidovski) [reply app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java:31]: Oh, I've missed that one. I have a couple of questions for this though: 
- What is this "Stellenantritt am"? On google, it translates it as "start of job on".
- Do we look at the month as a whole or the precise date. For example, if current date is 10.10.2020 and dat_stellenantritt_ab -> is this valid or no?

File: app-onlineform/src/test/java/ch/admin/seco/alv/service/onlineform/infrastructure/web/controller/AvpRestControllerIntTest.java (hunk 6 above)
Comment (mdavidovski) [reply app-onlineform/src/test/java/ch/admin/seco/alv/service/onlineform/infrastructure/web/controller/AvpRestControllerIntTest.java:80]: Because I also deleted the rest request for creating an AVP. I'm assuming we are not going to need it, since every new create AVP goest through the importer

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/UserInfo.java (hunk 13 above)
Comment (mdavidovski) [reply domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/UserInfo.java:13]: I was thinking of making it a common embeddable object for all entities that need user information, but I guess I can give it an Avp prefix now, and then extract it when we need to

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/AvpDto.java (hunk 9 above)
Comment (mdavidovski) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/dto/AvpDto.java:18]: sure, but I saw we need personNumber in the event dto, so I'll leave it in the AvpDto as well

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/AvpBusinessCaseImporterBatchConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    @Bean
+    Step closeAvpsHousekeepingStep() {
+		return stepBuilderFactory.get("close-old-avps-housekeeping-step")
+				.tasklet(this::avpCloseTasklet)
+				.build();
+    }
+
+	@Bean
+	@StepScope
+	JdbcCursorItemReader<StesInformation> lambdaActiveItemReader() {
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/AvpBusinessCaseImporterBatchConfiguration.java:101]: rename to something like `stesItemReader`

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/StesItemReaderFactory.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	private final AvpBusinessCaseImporterProperties avpBusinessCaseImporterProperties;
+
+	public StesItemReaderFactory(
+			@Qualifier("lamdaDataSource") DataSource lamdaDataSource,
+			AvpBusinessCaseImporterProperties avpBusinessCaseImporterProperties) {
+		this.lamdaDataSource = lamdaDataSource;
+		this.avpBusinessCaseImporterProperties = avpBusinessCaseImporterProperties;
+	}
+
+	JdbcCursorItemReader<StesInformation> lambdaActiveItemReader() {
```
Comment (emina-str) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/StesItemReaderFactory.java:44]: should we rename it to something like `stesItemReader` ?

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/AvpBusinessCaseImporterBatchConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+		return stesItemReaderFactory.lambdaActiveItemReader();
+	}
+
+	private AvpBusinessCaseEventProcessor avpBusinessCaseEventProcessor() {
+		return new AvpBusinessCaseEventProcessor(validator);
+	}
+
+	private RepeatStatus avpBusinessCaseDeleteTasklet(StepContribution contribution, ChunkContext chunkContext) {
+		Integer monthPast = avpBusinessCaseImporterProperties.getMonthPast();
```
Comment (cjpsycos) [inline app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/AvpBusinessCaseImporterBatchConfiguration.java:110]: why is it not mandatory (Nativ Int)? In properties is it. Please to it everywhere this property is used

File: app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java (hunk 3 above)
Comment (emina-str) [reply app-avp-business-case-importer/src/main/java/ch/admin/seco/alv/service/onlineform/config/LamdaBusinessCaseItemReaderFactory.java]: @mdavidovski I believe you have to look at the month as whole, but I will double check it with Ingo.
Because it is the same behaviour for NpA .. If this date is 10.10.2020. VOR-CP is 2020-10 and normal CP is only 2020-11 etc ..

So, basically for first implemtation just use the `now() >= dat_stellenantritt_ab` , and create an AvP for October even the unemployment date starts at 10.10.2020

**Review by emina-str — APPROVED** [review]
